File: Online_Problme_solving/Letter_Combinations_of_a_Phone_Number/l.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<string> letterCombinations(string digits) {
    if(digits.empty()) return vector<string>();
    vector<string> result;
    static const vector<string> v = {
		"", "", "abc", "def",
		"ghi", "jkl", "mno",
		"pqrs", "tuv", "wxyz"
	};
    result.push_back("");   
    for(int i = 0 ; i < digits.size(); ++i) {
        int num = digits[i]-'0';
        if(num < 0 || num > 9) break;
        const string& candidate = v[num];
        //if(candidate.empty()) continue;
        vector<string> tmp;
        for(int j = 0 ; j < candidate.size() ; ++j) {
            for(int k = 0 ; k < result.size() ; ++k) {
                tmp.push_back(result[k] + candidate[j]);
            }
        }
        result.swap(tmp);
    }
    return result;
}

int main() {
	string l;
	cin>>l;
	for(auto a : letterCombinations(l)){
		cout<<a<<endl;
	}
}

File: Online_Problme_solving/Letter_Combinations_of_a_Phone_Number/l.go
package main

import (
	"fmt"
)

func letterCombinations(digits string) []string {
	board := map[int]string{
		2: "abc", 3: "def",
		4: "ghi", 5: "jkl", 6: "mno",
		7: "pqrs", 8: "tuv", 9: "wxyz",
	}
	if len(digits) == 0 {
		return []string{}
	}
	ans := []string{""}
	for i := range digits {
		tans := []string{}
		num := int(digits[i] - '0')
		if num < 2 || num > 9 {
			break
		}
		c, e := board[num]
		if !e {
			continue
		}
		for _, s := range ans {
			for _, ss := range c {
				tans = append(tans, s+string(ss))
			}
		}
		ans = tans
	}

	return ans
}
func main() {
	fmt.Println(letterCombinations("233"))
}
